adapters/yandex/yandex.go

```diff
@@ -0,0 +1,311 @@
+		if len(value) > 0 {
+			headers.Add(key, value)
+		}
+	}
+}
+
+// Request is in shared memory, so we have to make a shallow copy for further modification (imp is already a shallow copy)
+func splitRequestDataByImp(request *openrtb2.BidRequest, imp openrtb2.Imp) openrtb2.BidRequest {
+	requestCopy := *request
+	requestCopy.Imp = []openrtb2.Imp{imp}
+
+	return requestCopy
+}
+
+func getYandexImpExt(imp openrtb2.Imp, yandexExt *openrtb_ext.ExtImpYandex) error {
+	var ext adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &ext); err != nil {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("imp %s: unable to unmarshal ext", imp.ID),
+		}
+	}
+
+	if err := json.Unmarshal(ext.Bidder, yandexExt); err != nil {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("imp %s: unable to unmarshal ext.bidder", imp.ID),
+		}
+	}
+
+	return nil
+}
+
+func modifyImp(imp *openrtb2.Imp) error {
+	if imp.Banner != nil {
+		banner, err := modifyBanner(*imp.Banner)
+		if banner != nil {
+			imp.Banner = banner
+		}
+		return err
+	}
+
+	if imp.Native != nil {
+		return nil
+	}
+
+	return &errortypes.BadInput{
+		Message: fmt.Sprintf("Unsupported format. Yandex only supports banner and native types. Ignoring imp id #%s", imp.ID),
+	}
+}
+
+func modifyBanner(banner openrtb2.Banner) (*openrtb2.Banner, error) {
+	height := banner.H
+	width := banner.W
+	format := banner.Format
+
+	if width == nil || height == nil || *width == 0 || *height == 0 {
+		if len(format) == 0 {
+			return nil, &errortypes.BadInput{
+				Message: "Invalid size provided for Banner",
+			}
+		}
+
+		firstFormat := format[0]
+		banner.H = &firstFormat.H
+		banner.W = &firstFormat.W
+	}
+
+	return &banner, nil
+}
+
+// "Un-templates" the endpoint by replacing macroses and adding the required query parameters
+func (a *adapter) resolveUrl(yandexExt openrtb_ext.ExtImpYandex, referer string, currency string) (string, error) {
+	publisherID := strconv.Itoa(int(yandexExt.PageID))
+	impID := strconv.Itoa(int(yandexExt.ImpID))
+	params := macros.EndpointTemplateParams{PublisherID: publisherID}
+
+	endpointStr, err := macros.ResolveMacros(a.endpoint, params)
+	if err != nil {
+		return "", err
+	}
+
+	parsedUrl, err := url.Parse(endpointStr)
+	if err != nil {
+		return "", err
+	}
+
+	addNonEmptyQueryParams(parsedUrl, map[string]string{
+		refererQueryKey:  referer,
+		currencyQueryKey: currency,
+		impIdQueryKey:    impID,
+	})
+
+	return parsedUrl.String(), nil
+}
+
+func addNonEmptyQueryParams(url *url.URL, queryMap map[string]string) {
+	query := url.Query()
+	for key, value := range queryMap {
+		if len(value) > 0 {
+			query.Add(key, value)
+		}
+	}
+
+	url.RawQuery = query.Encode()
+}
+
+func getReferer(request *openrtb2.BidRequest) string {
+	if request.Site == nil {
+		return ""
+	}
+
+	return request.Site.Domain
+}
+
+func getCurrency(request *openrtb2.BidRequest) string {
+	if len(request.Cur) == 0 {
+		return ""
+	}
+
+	return request.Cur[0]
+}
```

Comment:
```
        addNonEmptyQueryParams(parsedUrl, map[string]string{
                refererQueryKey:  referer,
                currencyQueryKey: currency,
                impIdQueryKey:    impID,
        })

        return parsedUrl.String(), nil
```

Adapter uses Referer and Currency as `addNonEmptyQueryParams`. However as per `getReferer` and `getCurrency` can be empty. @chernodub should return error if any one of these values is empty?

Reply:
Thanks a lot for noticing that @onkarvhanumante. The parameters were indeed intended as optional. The same logic works for our PBS Java adapter - https://github.com/prebid/prebid-server-java/blob/master/src/main/java/org/prebid/server/bidder/yandex/YandexBidder.java#L81